### ISSUE TITLE
Make ``Projection`` positional instead of label based

### ISSUE THREAD

**Issue by phofl** [dask/dask-expr#91]
I looked a bit into how an ``iloc`` implementation could look like (we need this for some of the more complicated reductions to adjust for potentially duplicated columns). I haven't gone done this rabbit hole terribly far, because I wanted to avoid doing something that you might have considered before.

Right now, ``Projection`` is label based, which works nicely but doesn't mesh very well with positional indexes. We could implement another class that does the same for positional column indexers, but this would require special casing every ``simplify_up`` call for both of them (simply because ``df[positional_indexer]`` does not work). If we'd convert the columns given to ``Projection`` into positional indexes, we could mesh both of them together and generalise the ``_simplify_up`` steps.

Also it makes estimating stuff easier, since label based won't tell you the dimensions of your Frame (duplicated columns), but this is a side-note.

cc @mrocklin @rjzamora wdyt?

**Comment by phofl** [comment]
Yeah duplicated columns get easier but aren't my main motivation (not that uncommon but also not super common I think). We'd get them for free if we make it positional.

Bigger motivation is that iloc and a getitem would expose the same interface. The only difference would be that ``Projection`` would convert it's columns into a positional indexer, which is what we would expose to all other operations. 

We can also punt on iloc for now, but I am a bit concerned that we'd have to change that stuff later on.

**Comment by mrocklin** [comment]
> Bigger motivation is that iloc and a getitem would expose the same interface

I'm not sure I understand what you mean here.  Can you say more?

> The only difference would be that Projection would convert it's columns into a positional indexer, which is what we would expose to all other operations

I think that semantically I prefer storing column names rather than column positions.  This feels closer to user intent to me.

**Comment by mrocklin** [comment]
> We can also punt on iloc for now

In general I support this, unless it's going to slow us down on other work.  

> but I am a bit concerned that we'd have to change that stuff later on.

If it's much more expensive to do this later then yes, I'd agree.  I'm not currently seeing anything concerning though.  I guess maybe if we have to shift all projection optimizations to use positions rather than column names that would be somewhat annoying.  I'm not sure about that shift though (as I said I think that we may prefer names in the future) and it doesn't sound that expensive of a move to make if we need to.

**Comment by phofl** [comment]
> I'm not sure I understand what you mean here. Can you say more?

If your ``parent`` is a Projection, we are generally doing something like:

```
self.frame[parent.operand("columns")]
```

while with positional indexers we would have to do

```
self.frame.iloc[:, parent.operand("position_indexer")]
```

Unifying both of them would be convenient since we don't have to special case for both. We can do this both ways though, converting positional indexers to label based ones or label base indexers to positional ones.

> I think that semantically I prefer storing column names rather than column positions. This feels closer to user intent to me.

If we say that we don't care about duplicate columns then we can also go the other way and convert positional indexes from iloc into label based values like we store in Projection, this would work as well.

We are using iloc under the hood a couple of times, this might become a problem for var, std and describe.
